### ISSUE TITLE
[Feature Request] Move automations, scripts and helper entities to internal integration functions

### ISSUE THREAD

**Issue by Snuffy2** [FutureTense/keymaster#393]
From @raman325 https://github.com/FutureTense/keymaster/issues/334#issuecomment-1566588123:
> You are right that `keymaster` has some unnecessary complexity/surface area, but by comparing it to this script you are ignoring a lot of what keymaster does. Ultimately, I'd like keymaster to be more modular because a lot of people don't need all of the features that keymaster offers, but that's a significant rewrite and would not be made simpler by moving to pyscript or App Daemon.
> If this is something you feel that confidently about, feel free to submit a PR for this or a fork.
> 
> There is no design flaw. At the time that keymaster was originally created, the implementation was the only solution given what was available. There's a path to a better solution but that's only been made available recently, and that path does not involve pyscript or App Daemon.
>
> Great - have at it and please share once it's built!

I'm wondering if there would be interest in a rewrite of sorts moving away from automations, scripts, and helpers? I've started thinking about how best to do it and I am interested in giving it a try if there would be interest from the repo maintainers in this.

I'm **very** open to feedback but my approach would be:
* Create a device for each lock in keymaster with all of the keymaster entities for it
  * For child locks, set the `via_device` property to the parent lock
* Expand the KeymasterLock class to store all of the properties of the lock and utilize the device_ids to link parent/child locks
* Utilize a coordinator to hold all the locks so that each lock can access and change the properties of its parent/child locks
* Move automations/scripts to internal integration functions
* Move from helper entities to integration entities

| **Helper Entity** | **Integration Entity** |
|-------------------|------------------------|
| input_text | text |
| input_boolean | switch |
| input_number | number |
| input_datetime | date, datetime, or time |

Obviously this would not be a single, small PR. 

Before I go further down this rabbit-hole, please let me know if there's interest. If not, no worries.

**Comment by firstof9** [comment]
Sounds good to me.

**Comment by bergethan** [comment]
I'd love to see this! Because of how the automations are generated, I cannot categorize them using Home Assistant's new categorization features, so I have to scroll endlessly to find anything else I'm looking for. It's not sustainable long-term.

**Comment by firstof9** [comment]
@Snuffy2 give me a ping when you're ready for those to be merged into that beta branch.

**Comment by tykeal** [comment]
Looks like I'm going to have to install a beta version on my test system to make sure that Rental Control still works with all of this! Wee!

I'll note that all of those PRs are currently failing the py312 tests and that really should be fixed before they're merged anywhere.

Linting is also failing, but I suspect that's symptomatic of the entire code base. I haven't checked the current code base against the linters ;)

**Comment by firstof9** [comment]
> I'll note that all of those PRs are currently failing the py312 tests and that really should be fixed before they're merged anywhere.

The tests are broken due to internal changes in HA. They need reworking, I just don't have the time for such the endeavor.

**Comment by Snuffy2** [comment]
@firstof9 I think this is now ready for beta testing.

It utilizes all internal code to manage keymaster with the exception that it will continue to send notices to `keymaster_<doorname>_manual_notify` if it exists.
I tried to reproduce the same logic as before but I'm sure there are some differences. Let me know and I'll try to refine it
I've tested it locally but it really needs some more extensive testing, hence the beta

#### Prior to release, I'll remove a lot of the logging and remove the actual PIN/Usercodes from the logs.

## Functionality not yet implemented
* A _best effort_ migration function to convert from the helper entities to internal entities
* Remove prior keymaster package files and entities in the migration
* Lovelace code generator
* HA actions/scripts
* Add button entity to reset (lock and each code slot)

## Features
* Coordinator redesigned to manage the status of all keymaster locks and the link to the physical locks
  * Entities now get their value/state from the Coordinator
* Expand the KeymasterLock class to hold all of the necessary lock properties
* Standardize the code of entities using a KeymasterEntity base class
* Converted existing Sensor and Binary Sensors
* Added Text entities replacing input_text helper entities
* Create a device for each lock in keymaster with all of the keymaster entities for it
  * For child locks, set the via_device property to link the parent lock to the child lock
* Move input_boolean helper entities to switch entities
* Move input_number to number entities
* Move input_datetime to datetime or time entities
* Store coordinator data to disk and restore on HA restart
* Build Code Slot active logic
* Move applicable automations and scripts to listeners
* Add boltchecked/retry lock function
* Add auto-lock timer function
* Ensure new and change configurations work properly

**Comment by firstof9** [comment]
Now we just need to fix tests.

**Comment by Snuffy2** [comment]
> Now we just need to fix tests.

Not something I have a lot of experience with but I'll give it a try. Any assistance would be greatly appreciated.

**Comment by firstof9** [comment]
I'll have to see what I can do when time permits.

**Comment by tykeal** [comment]
@Snuffy2 I've cleaned up my old keymaster install on my test system and have installed the beta branch to walk through some stuff.

First up, I saved off one of the slot config boxes from my previous install and am working through re-attaching the different entities.

There are two entities that I have yet to find a replacement for yet. The connected sensor and the active sensor. What are the new entities that I should be pointing at for this?

Second: setting the PIN doesn't seem to actually cause my lock to get programmed, or deprogrammed. Flipping enabled / disabled doesn't seem to be doing it either, all it seems to do is set the PIN to unknown when it goes from disabled -> enabled.

Third: if the slot is not enabled it is impossible to program _any_ of the entities which is a major problem for when I update Rental Control to be able to use this. Rental Control programs the name, the pin, the start / stop daterange and that it should be using the date range all before enabling the slot. That way the slot doesn't try push a code out when it isn't supposed to.

If the logic is to enable the slot, then program the advanced options, and _then_ program the name and PIN I can do that but it's a reworking of logic from how keymaster currently works.

**Comment by Snuffy2** [comment]
Thanks for testing it out.

1. I have not recreated the code slot sensor called `sensor.connected_<lockname>_<#>`. With the new setup, I was hoping that it wouldn't be needed but I can add it in if necessary.
* For Active per Code Slot: There should be a binary_sensor called something like `binary_sensor.<lock_name>_code_slot_<#>_active`.
* For Lock Connected: There should be a binary sensor called something like: `binary_sensor.<lockname>_network`.

2. Could you enable debug logging and either post them or email them to me (github@mail.snuffy2.com). __Note:__ that the logs currently contain PIN/usercodes in them while testing/in beta. Feel free to de-identify but note if they are discrepant between keymaster and the lock please.

3. Yes, that is how I built it as that seemed to make sense. But I can change it to function how it used to work since there is a valid reason/use case as to why you'd want to set a pin even when disabled.

**Comment by tykeal** [comment]
I'll turn on debugging to get some output, On the topic of the connected and active sensors. The connected sensor is the most critical as that has been my way of knowing if the PIN has actually hit the lock without having to go inspect it directly in ZWave-JS UI. We need _some_ sort of indicator that the PIN has been properly programmed or if there is a problem with it. The active sensor just lets you know what state that the PIN _should_ be in. It allows you to know when Keymaster should be programming or deprogramming. But in reality the connected sensor tells you all of that with the states that it has of `Disconnected`, `Adding`, `Connected`, and `Deleting`.

Having at least the connected sensor is rather vital as a piece of knowledge as to what the integration is doing. I have one lock that fairly consistently runs into that `Adding` hang that is all over the issue tracker here and I just go fix the given code slot in JWave-JS UI and make sure it's properly programmed and then it shows as `Connected` in Keymaster. I don't have that issue with the many, many other locks I have in use but they do all sometimes run into it and it's useful for me to know what Keymaster is thinking.

------ A little while later -----

Ok, I enabled debugging and here's what I've found:

1.  I initially setup the new configuration to work on the first 10 slots of my test lock. I had codes programmed into slot 1 and 3 with `0101` and `0303` respectively.

1. The logs are showing me both of those values.

1. I later went and changed the configuration to only work on the first slot.

1. Restarts of the system still show the coordinator with 10 slots, even after a full restart of HA

1. Complete removal of current configuration + restart + adding it again still has it showing up with 10 slots

1. Slot programming was not working correctly _until_ I manually set the PIN in ZWave-JS UI on the first test slot. Since then it's been doing the correct thing with programming / deprogramming the slot. I see that there is a `binary_sensor.<lock_name_code_slot_<#>_active` that seems to be effectively doing what I expected the end state status of the `connected` sensor to do. It does not, however tell me if Keymaster is trying to program / deprogram the slot. Granted, the lock I'm testing with rarely has had the hung bug that seems so prevalent so I don't know if it matters here.

1. Slot 3 is an issue though. I deprogrammed the slot manually via ZWave-JS UI and I still see it showing up in the log with the PIN that it had long previously detected when I had it doing 10 slots.

1. Removing a PIN code (setting it to empty) throws the following error:

```
2024-11-26 08:14:17.355 DEBUG (MainThread) [custom_components.keymaster.text] [Text async_set_value] test_lock Code Slot 1: PIN: config_entry_id: 01JDMEKFV57AHZSSG3GDD6YH3Q, value:
2024-11-26 08:14:17.355 DEBUG (MainThread) [custom_components.keymaster.coordinator] [set_pin_on_lock] test_lock: Code Slot 1: Setting PIN to
2024-11-26 08:14:17.355 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [140006301334496] Unexpected exception
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 245, in handle_call_service
    response = await hass.services.async_call(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2802, in async_call
    response_data = await coro
                    ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2845, in _execute_service
    return await target(service_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 1007, in entity_service_call
    single_response = await _handle_entity_call(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 1079, in _handle_entity_call
    result = await task
             ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/text/__init__.py", line 81, in _async_set_value
    await entity.async_set_value(value)
  File "/config/custom_components/keymaster/text.py", line 119, in async_set_value
    await self.coordinator.set_pin_on_lock(
  File "/config/custom_components/keymaster/coordinator.py", line 1282, in set_pin_on_lock
    await set_usercode(kmlock.zwave_js_lock_node, code_slot, pin)
  File "/usr/local/lib/python3.12/site-packages/zwave_js_server/util/lock.py", line 146, in set_usercode
    raise ValueError("User code must be at least 4 digits")
ValueError: User code must be at least 4 digits
```

There should be a way to fully clear the code outside of just disabling it. Again, this comes back to not being able to program the code while the slot is disabled. Changing out a code needs to happen without an old code hitting the lock temporarily.

9. Adding a new configuration after removing an old one ends up requiring a reload of the integration or restart of HA to get it to configure the entities, they all end up as `Unavailable` without doing this and the following gets dumped in the log:

```
2024-11-26 08:30:56.622 DEBUG (MainThread) [custom_components.keymaster.coordinator] [add_lock] test_lock
2024-11-26 08:30:56.622 DEBUG (MainThread) [custom_components.keymaster.coordinator] [update_door_and_lock_state] Running
2024-11-26 08:30:56.622 DEBUG (MainThread) [custom_components.keymaster.coordinator] [unsubscribe_listeners] test_lock: Removing all listeners
2024-11-26 08:30:56.622 DEBUG (MainThread) [custom_components.keymaster.coordinator] [update_listeners] test_lock: Calling create_listeners now
2024-11-26 08:30:56.622 DEBUG (MainThread) [custom_components.keymaster.coordinator] [create_listeners] test_lock: Creating handle_zwave_js_lock_event listener
2024-11-26 08:30:56.623 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.624 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.625 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.625 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.626 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.627 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.627 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.631 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.633 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.633 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.633 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.634 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.634 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.634 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.635 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.635 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.635 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.635 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.636 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.637 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.638 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.639 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.639 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.640 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.641 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.641 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.641 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.642 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.647 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.647 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.648 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.648 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.649 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.649 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.650 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.650 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.651 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.651 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.654 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.655 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.656 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.656 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.657 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.657 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.657 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.658 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.658 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.658 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.658 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.659 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.661 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.661 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.661 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
2024-11-26 08:30:56.661 DEBUG (MainThread) [custom_components.keymaster.coordinator] Debouncer call ignored as shutdown has been requested.
```

While not exactly common, if someone messes up and wants to restart their config from scratch this is likely something that they would end up doing.

**Comment by Snuffy2** [comment]
@tykeal, hopefully this will work a bit better now.

* #404 should fix numbers 1-3, 6-8. 
* I added a Sync Status sensor and updated a lot of the surrounding logic and logging
* You can now change the code slot settings on disabled code slots
* I'm not able to reproduce the issues you noted in numbers 4 & 5. It consistently only shows the number of slots I select. Let me know if you're still seeing this and anything of note you see in the debug logs.
* Number 9, I see on occasion when there's another error in the logs. I'm not quite clear on how to address the Debouncer issues directly but fixing the issue that caused the error makes it so the Debouncer issues don't appear. Let me know if you're still seeing it and if there's a keymaster error higher up in the logs

So once #404 gets merged, if you could test again I'd really appreciate it. Also, there is now an upgrade/merge function added that should keep the settings from the previous version and delete the old entities when you upgrade from the current version to the beta.

**Comment by firstof9** [comment]
#404 is merged.

**Comment by Snuffy2** [comment]
I've updated the Wiki based on the changes currently in the beta for when this gets fully released.
https://github.com/Snuffy2/keymaster-wiki/wiki
https://github.com/Snuffy2/keymaster-wiki.wiki.git

**Comment by tykeal** [comment]
I'll try and get some testing done soon. I was away for Thanksgiving and have a busy weekend currently in front of me.

As a point of clarification, I should note that when I say "Restarts of the system still show the coordinator with 10 slots, even after a full restart of HA" I mean that it is informing me of information about slot states, not actually creating entities for 10 slots. I haven't pushed the latest beta update but I did finally notice this log line:

```2024-11-26 08:33:15.748 DEBUG (SyncWorker_5) [custom_components.keymaster.coordinator] [Coordinator] No changes to kmlocks. Not updating JSON file```

which made me go searching for a json file, I was originally thinking it would have been in `~/config/.storage` with all the HA configuration but it isn't, and I think I realize what is happening. Modifications to the configuration are not clearing out old slots from the `config/custom_components/keymaster/json_kmlocks/keymaster_kmlocks.json` file. Given the number of times I've added and removed the lock while doing some testing I now have 5 blocks in the file.

The first block is my initial just install it over the old Keymaster without removing configuration and see what happens. I had that KM configured to do slots 10 - 14. When that didn't work very well, I did a complete removal of any configuration and built new that had slots 1 - 10. Since that was too much clutter while I was trying to figure out the updated entities for the dashboard I _modified_ the configuration to only be slot 1. I have since removed all configuration, restarted HA and added configuration for only 1 slot 2 other times.

This tells me that on removal you are not clearing your data blocks at all. On a reconfiguration, you're leaving the data block behind and creating a new one instead of updating the old one. And finally, data from all blocks seem to be getting reported on the regular check cycle which is what is was confusing me. I will try purging that file completely after removing all of my configurations and giving HA a restart to see if that fixes my problem.

**Comment by tykeal** [comment]
Ok, after some further testing with the latest beta branch _and_ having purged the json file things are definitely looking a lot better! I like the sync status entity. Especially when I just remove the PIN (thanks for cleaning up the error that was thrown). Since it goes to `Out of Sync` that's really good :)

I have also loaded up the generated dashboard and that looks good. Even better than the previously generated one since all the advanced features are truly hidden unless enabled. That's a good touch :) I will note, however, that you've moved the location of the generated dashboard. The documentation will need to be updated _or_ you should take a look at how [Lock Code Manager](https://github.com/raman325/lock_code_manager) by @raman325 does an autogenerated dashboard. I don't really understand how it works, but it's pretty slick when all you have to do is add a dashboard and include a single custom card and all the locks and everything just auto-populate!

A reset button is still needed however, that will make it significantly easier for someone to clear out slot along with all of the advanced configuration. I use this feature of the current release with my Rental Control integration so that I don't have to explicitly clear the name, pin, enabled status and date range when a code slot needs to be cleared.

**Comment by tykeal** [comment]
@Snuffy2 I haven't dug into the code to see what is going on, but when I'm using my programmed code to unlock the lock, it's firing a `keymaster_lock_state_changed` event (like it should) but it's throwing it as `code_slot: 0` with `code_slot_name: ""` instead of the actual slot information

Since I only have one slot currently defined, it should be throwing it as `code_slot: 1` with `code_slot_name: "foo"` Admittedly, it could actually be the lock. I had pulled this one from one of my properties because it was regularly having issues... I do not presently have a verified good lock to test with :rofl:

**Comment by Snuffy2** [comment]
@tykeal 
* Deleting locks should already be working better, but I also made some further improvements with #409
* Reset lock and code slots added with #408 
* Updated documentation has been started for whenever this gets released: https://github.com/Snuffy2/keymaster-wiki/wiki
* I'll look into the notifications. It isn't something I tested very thoroughly yet, so I'm not shocked if it isn't working right.
* Interesting approach to the dashboard with Lock Code Manager. Not sure I plan to deploy anything similar right now, but an approach to consider for this (or other integrations) in the future

@tykeal, @firstof9 do you think this is in a good enough state where we can try to get some more users to test it out? Any ideas how to recruit some users?

**Comment by tykeal** [comment]
I'll do some more testing when I can get a chance either this evening or in the morning.

As for more users testing, it could go out as a beta 1.0 this is a major breaking change and I would posit closer to what a 1.0 really should be than what we've had before.

**Comment by bergethan** [comment]
@Snuffy2, I have a setup with several Yale locks. Since the work you’re doing is something I’ve specifically asked for (and I’m REALLY looking forward to), I’m happy to test it out.

By the way, in case nobody has said it yet, thank you!

**Comment by tykeal** [comment]
Heya, just touching base. I know I said I would do some more testing, but I haven't had a chance to (life has intruded). I'll try as soon as I can!

**Comment by Snuffy2** [comment]
@firstof9 I have the changes to pass mypy just about done and I have one other change regarding the notifications to do but otherwise that's it.

I've tried getting pytest working, but I'm out of my depth in how to do it properly with the coordinator and zwave, sorry.

What are next steps to get this to a wider beta audience?

**Comment by firstof9** [comment]
I'll check in on the pytest to see what's going on, likely just needs some tweaking for the changes to zwave_js that have been done in the past few years.

**Comment by Snuffy2** [comment]
With #428, I am now "done". Without any further testing by others, I'm don't anticipate finding any other issues. 

@firstof9, please let me know what next steps are to get this further tested and hopefully released.

**Comment by CJOWood** [comment]
> With #428, I am now "done". Without any further testing by others, I'm don't anticipate finding any other issues. 
> 
> @firstof9, please let me know what next steps are to get this further tested and hopefully released.

I'm curious if you can explain some of the changes more clearly between this and the original. As far as I understand it's putting some of the logic being scripts rather than in helpers?

**Comment by Snuffy2** [comment]
It uses all internal functions. It doesn't use the packages, helpers, automations, scripts that the current version uses. The only script/action it will optionally use is a custom notification script/action. I believe it has all of the options and functionality of the current version.

**Comment by tykeal** [comment]
@Snuffy2 this changes entity names as well too doesn't it? I should finally have the breathing space to do more testing with it today. Sorry, from Thanksgiving until after the New Year is always extremely full in my family so I haven't been able to do the needed testing.

**Comment by VACInc** [comment]
What's the best way to test this out without nuking the existing?  I would assume fork it and install it that way?  Would love to also verify functionality with my 4 (1 leader, 3 follower) Schlage locks

Also down to just test out upgrade from my current keymaster instance if migration is present now

**Comment by firstof9** [comment]
> What's the best way to test this out without nuking the existing?

I spin up a 2nd HA container with a minimal config and do it that way.

**Comment by Snuffy2** [comment]
Yes, entities will change. Moving from all helper entities to integration-based entities necessitates that.

I have built a migration from the current keymaster to the new one. It works for me but it all needs more testing to ensure it fully works.

**Comment by VACInc** [comment]
I'll take a backup and live on the edge (give it a go) when I get a second this week

**Comment by tykeal** [comment]
> Yes, entities will change. Moving from all helper entities to integration-based entities necessitates that.
> 
> I have built a migration from the current keymaster to the new one. It works for me but it all needs more testing to ensure it fully works.

Which will be fine for everyone that is using Keymaster directly. Folks that are using it as a dependency of [Rental Control](https://github.com/tykeal/homeassistant-rental-control) are going to need to wait until I get that updated to handle the new entities.

**Comment by CJOWood** [comment]
> Yes, entities will change. Moving from all helper entities to integration-based entities necessitates that.
> 
> I have built a migration from the current keymaster to the new one. It works for me but it all needs more testing to ensure it fully works.

I'd be happy to test it out. I'm planning on giving it all a test run today or tomorrow.

**Comment by CJOWood** [comment]
> Yes, entities will change. Moving from all helper entities to integration-based entities necessitates that.
> 
> I have built a migration from the current keymaster to the new one. It works for me but it all needs more testing to ensure it fully works.

@Snuffy2 I would be happy to test this out on my setup (1 master lock with 2 child and 20 slots each). Would love to try it out tonight or tomorrow.

**Comment by Snuffy2** [comment]
For anyone willing to test, for now, you will need to manually "install" the beta.
* Do not uninstall your current keymaster
* Download it from https://github.com/FutureTense/keymaster/archive/refs/heads/beta.zip
* Extract it
* Move all the files in `keymaster-beta/custom_components/keymaster` to your Home Assistant `custom_components/keymaster` directory. Overwrite the existing files.
* Edit your HA `configuration.yaml` to enable debug logging for keymaster
* Restart Home Assistant


To enable debug logging, add this or modify the logging section of your Home Assistant `configuration.yaml`:
```yaml
logger:
  default: warning
  logs:
    custom_components.keymaster: debug
```

**Comment by CJOWood** [comment]
> For anyone willing to test, for now, you will need to manually "install" the beta.
> * Download it from https://github.com/FutureTense/keymaster/archive/refs/heads/beta.zip
> * Extract it
> * Move all the files in `keymaster-beta/custom_components/keymaster` to your Home Assistant `custom_components/keymaster` directory
> * Restart Home Assistant
> 

@Snuffy2 Is this with or without uninstalling the current version? Was there a update script we can test?

**Comment by Snuffy2** [comment]
Don't uninstall, just overwrite the existing files.

**Comment by CJOWood** [comment]
Removed a lock to test as I had some Lovaelace issues. I am getting this error when I try to re-add.

![image](https://github.com/user-attachments/assets/ca951894-243a-4101-b113-3f6478f97bbb)

In the log it looks like it clears all the slots then adds the set codes back before throwing this error:
`2025-01-10 20:31:06.295 ERROR (MainThread) [custom_components.keymaster] Timeout on add_lock. CancelledError:`

kmlocks.json is showing a 'garagedoor' entry even though it was deleted.

**Comment by CJOWood** [comment]
Okay, looks like it was an issue with the `kmlock.json` file. I edited the `garagedoor` entry (added _old) and I am now able to add it as a lock again... might have messed up my `frontdoor` lock though :p

**Edit: ** When (now) attempting to remove the old entry the integration fails to start. I have to keep the old garagedoor_old entry in the JSON file or it fails to start.

**Edit2: ** There was a reference to the child lock in the parent lock's entry... I removed that and it worked.

**Comment by CJOWood** [comment]
@Snuffy2 Some really funky things are happening. It seems to be constantly clearing and setting codes on my 2 child locks. I have a Schlage BE469ZP and a Kwikiset SmartCode 888 and never end up being enabled.

**Comment by Snuffy2** [comment]
@CJOWood, can you please enable debug logging and post some of the logs when you are having issues? I'm having trouble sorting out what the issues are.

To enable debug logging, add this or modify the logging section of your Home Assistant `configuration.yaml`:
```yaml
logger:
  default: warning
  logs:
    custom_components.keymaster: debug
```

**Comment by CJOWood** [comment]
[HA Keymaster Beta Log](https://github.com/user-attachments/files/18387105/home-assistant_2025-01-11T19-32-52.009Z.log)

Here is the log from Keymaster. On line ~2108 I disabled the second child lock (`garagedoor`) to simplify things. My main door (`frontdoor`) and the `garagedoor` lock are the same (Schlage BE469ZP). `backdoor` is the Kwikiset 888. `frontdoor` seems to be working fine.

I tried disabling and re-enabling, as well as re-configuring the backdoor as a parent lock rather than child. Line `2600` is where the cycling is very clear.

At line `2953` there is some cycling of pins for `frontdoor`. Code slot #3 is not enabled.

~~I **vaguely** remember in very early Keymaster development that we had a similar cycling issue... @firstof9 @raman325  Do you guys remember that? Did it have to do with the zwave action call? Or maybe the setting some locks to `0000` rather than `null`?~~ **Nevermind**, it (#117) was just an issue with clearing - which this definitely does not have a problem with!

**Edit:** I also just tried to exclude and re-include the `backdoor` lock and then re-add to Keymaster. Still have the same issue.

**Comment by muddro1** [comment]
Just came across this and really excited to see the update here. Will check out the beta on my system and let you know if any issues. nly have 1 lock so not a complex setup, and if it works Ill add to my folks set up which has 4 locks (1 parent, 3 children). Curious, is this also going to "modernize" the dashboard with sections and current iteration of badges?

**Comment by Snuffy2** [comment]
@CJOWood it looks like something got messed up when I made a bunch of changes to comply with mypy. I think I have much of it fixed and hope to have a PR out to the beta later today to get back to where it was before.

**Comment by CJOWood** [comment]
> @CJOWood it looks like something got messed up when I made a bunch of changes to comply with mypy. I think I have much of it fixed and hope to have a PR out to the beta later today to get back to where it was before.

Can I try a version on your fork?


**Found a version. Here is a new log.** Some issue with always clearing code slot 0?
```
2025-01-13 21:10:24.081 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] frontdoor: Code Slot 3: Clearing PIN
2025-01-13 21:10:25.893 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] frontdoor: Code Slot 3: Clear command sent, confirming
2025-01-13 21:10:25.893 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] frontdoor: Code Slot 3: PIN Cleared
2025-01-13 21:10:41.721 DEBUG (MainThread) [custom_components.keymaster.coordinator] [handle_zwave_js_lock_event] backdoor: event: <Event zwave_js_notification[L]: domain=zwave_js, node_id=31, home_id=4155999415, endpoint=0, device_id=147911c3828b2078b2afc4194581497b, command_class=113, command_class_name=Notification, label=Access Control, type=6, event=13, event_label=Single user code deleted, parameters=nodeId=31, ccId=Z-Wave Long Range>, new_state: unlocked, params: {'nodeId': 31, 'ccId': 'Z-Wave Long Range'}, code_slot_num: 0
2025-01-13 21:10:53.144 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 5: Clear command sent, confirming
2025-01-13 21:10:53.145 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 5: PIN Not Cleared, will retry
2025-01-13 21:10:53.146 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 6: Clearing PIN
2025-01-13 21:10:53.358 DEBUG (MainThread) [custom_components.keymaster.coordinator] [handle_zwave_js_lock_event] backdoor: event: <Event zwave_js_notification[L]: domain=zwave_js, node_id=31, home_id=4155999415, endpoint=0, device_id=147911c3828b2078b2afc4194581497b, command_class=113, command_class_name=Notification, label=Access Control, type=6, event=13, event_label=Single user code deleted, parameters=nodeId=31, ccId=0x05>, new_state: unlocked, params: {'nodeId': 31, 'ccId': '0x05'}, code_slot_num: 0
2025-01-13 21:11:04.697 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 6: Clear command sent, confirming
2025-01-13 21:11:04.697 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 6: PIN Not Cleared, will retry
2025-01-13 21:11:04.697 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 8: Clearing PIN
2025-01-13 21:11:04.908 DEBUG (MainThread) [custom_components.keymaster.coordinator] [handle_zwave_js_lock_event] backdoor: event: <Event zwave_js_notification[L]: domain=zwave_js, node_id=31, home_id=4155999415, endpoint=0, device_id=147911c3828b2078b2afc4194581497b, command_class=113, command_class_name=Notification, label=Access Control, type=6, event=13, event_label=Single user code deleted, parameters=nodeId=31, ccId=0x06>, new_state: unlocked, params: {'nodeId': 31, 'ccId': '0x06'}, code_slot_num: 0
2025-01-13 21:11:16.304 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 8: Clear command sent, confirming
2025-01-13 21:11:16.305 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 8: PIN Not Cleared, will retry
2025-01-13 21:11:16.305 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 10: Clearing PIN
2025-01-13 21:11:16.544 DEBUG (MainThread) [custom_components.keymaster.coordinator] [handle_zwave_js_lock_event] backdoor: event: <Event zwave_js_notification[L]: domain=zwave_js, node_id=31, home_id=4155999415, endpoint=0, device_id=147911c3828b2078b2afc4194581497b, command_class=113, command_class_name=Notification, label=Access Control, type=6, event=13, event_label=Single user code deleted, parameters=nodeId=31, ccId=0x08>, new_state: unlocked, params: {'nodeId': 31, 'ccId': '0x08'}, code_slot_num: 0
2025-01-13 21:11:27.933 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 10: Clear command sent, confirming
2025-01-13 21:11:27.933 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 10: PIN Not Cleared, will retry
2025-01-13 21:11:27.933 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 11: Clearing PIN
2025-01-13 21:11:28.154 DEBUG (MainThread) [custom_components.keymaster.coordinator] [handle_zwave_js_lock_event] backdoor: event: <Event zwave_js_notification[L]: domain=zwave_js, node_id=31, home_id=4155999415, endpoint=0, device_id=147911c3828b2078b2afc4194581497b, command_class=113, command_class_name=Notification, label=Access Control, type=6, event=13, event_label=Single user code deleted, parameters=nodeId=31, ccId=0x0a>, new_state: unlocked, params: {'nodeId': 31, 'ccId': '0x0a'}, code_slot_num: 0
2025-01-13 21:11:39.614 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 11: Clear command sent, confirming
2025-01-13 21:11:39.614 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 11: PIN Not Cleared, will retry
2025-01-13 21:11:39.614 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 12: Clearing PIN
2025-01-13 21:11:39.777 DEBUG (MainThread) [custom_components.keymaster.coordinator] [handle_zwave_js_lock_event] backdoor: event: <Event zwave_js_notification[L]: domain=zwave_js, node_id=31, home_id=4155999415, endpoint=0, device_id=147911c3828b2078b2afc4194581497b, command_class=113, command_class_name=Notification, label=Access Control, type=6, event=13, event_label=Single user code deleted, parameters=nodeId=31, ccId=0x0b>, new_state: unlocked, params: {'nodeId': 31, 'ccId': '0x0b'}, code_slot_num: 0
2025-01-13 21:11:51.115 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 12: Clear command sent, confirming
2025-01-13 21:11:51.115 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 12: PIN Not Cleared, will retry
2025-01-13 21:11:51.116 DEBUG (MainThread) [custom_components.keymaster.coordinator] [clear_pin_from_lock] backdoor: Code Slot 1: Clearing PIN
2025-01-13 21:11:51.358 DEBUG (MainThread) [custom_components.keymaster.coordinator] [handle_zwave_js_lock_event] backdoor: event: <Event zwave_js_notification[L]: domain=zwave_js, node_id=31, home_id=4155999415, endpoint=0, device_id=147911c3828b2078b2afc4194581497b, command_class=113, command_class_name=Notification, label=Access Control, type=6, event=13, event_label=Single user code deleted, parameters=nodeId=31, ccId=0x0c>, new_state: unlocked, params: {'nodeId': 31, 'ccId': '0x0c'}, code_slot_num: 0
```

**Comment by Snuffy2** [comment]
The PR has been merged and its in the beta branch here now, but I see you found it on my fork too.

I see the same thing on occasion in my logs when deleting a code slot. It is continuously retrying as the code is just not getting cleared for some reason. I send the clear_usercode command and the function doesn't return an error and ZwaveJS fires an event saying it is cleared. However, when I recheck the usercodes the cleared code is still there. Thus, the code continues to try until it actually clears. It generally will clear the code after a few minutes. I don't have the issue when adding or changing a code, just clearing. 

Note that I see the same thing when I do it manually using the Home Assistant service (action) `zwave_js.clear_lock_usercode`. It takes multiple tries before the code slot will clear. 

The [handle_zwave_js_lock_event] showing `code_slot_num: 0` at the end doesn't actually mean its related to code slot 0 (I'll clean up the logging), it just means that there wasn't a code slot in the event details.

**Comment by muddro1** [comment]
@Snuffy2 , not sure if this is purposeful, but all the entities create some gibberish in their entity names. For example: switch.keymaster_01jhdf43frstpkkd2zrfkc0fqe_switch_autolock_enabled. Is there a way to use the lock name or something instead of the 01....?

EDIT: NM. I removed the integration and readded and seems to have clean entity names now.

**Comment by tykeal** [comment]
@Snuffy2 finally getting around to more testing. So far things look good and I've found a way to auto-detect if Keymaster has been upgraded to this version or not so I can make Rental Control compatible with both the current release and a future release with this.

That being said, I see something missing that would be really nice to have. A `Reset` button or toggle like the current release has to reset a slot completely. I use this with Rental Control for managing the slots and also for the few slots that I manually control if I have some that are programmed for long term guests that aren't going through my calendars for whatever reason.

I haven't tested how well pre-programming everything on the slot works before enabling it, but I believe I saw that you made it possible to do that unlike before.

**Comment by tykeal** [comment]
Correction, it does exist, but it's not being added to the generated dashboard template by default. That's fine! I'll end up using a custom dashboard anyway, just as long as it exists 😄

**Comment by tykeal** [comment]
@Snuffy2 so, I'm trying to get Rental Control able to work with the update but I'm running into a major issue related to the decision to keep the advanced options disabled until a switch is set. If those entities were just hidden from the dashboard instead of being disabled if the switch is off (in particular the `switch.{lockname}_code_slot_{number}_use_date_range_limits`) because of the way the start & end entities get enabled / reset when the reset button is used it's causing some major problems with my integration since I have to track what is stored in case the end user makes certain types of changes. Would you consider a change to making all of the entities always enabled and they just get hidden / do nothing if the relevant switch is turned off?

**Comment by jleach87** [comment]
I've got an issue with a Yale lock. I've tried using the keymaster switches to disable and re-enable the slots. I've also tried setting in the Z-Wave JS UI to no avail. Seems to be trying every few seconds.

```
Logger: custom_components.keymaster.coordinator
Source: custom_components/keymaster/coordinator.py:1404
integration: keymaster ([documentation](https://github.com/FutureTense/keymaster), [issues](https://github.com/FutureTense/keymaster/issues))
First occurred: 12:43:48 AM (4772 occurrences)
Last logged: 10:48:57 AM

    [Coordinator] front_door: Code Slot 1: Unable to clear PIN. NotConnected:
    [Coordinator] front_door: Code Slot 3: Unable to clear PIN. NotConnected:
    [Coordinator] frontdoor: Code Slot 1: Unable to clear PIN. NotConnected:
    [Coordinator] frontdoor: Code Slot 3: Unable to clear PIN. NotConnected:
```

**Comment by tykeal** [comment]
> I've got an issue with a Yale lock. I've tried using the keymaster switches to disable and re-enable the slots. I've also tried setting in the Z-Wave JS UI to no avail. Seems to be trying every few seconds.
> 
> ```
> Logger: custom_components.keymaster.coordinator
> Source: custom_components/keymaster/coordinator.py:1404
> integration: keymaster ([documentation](https://github.com/FutureTense/keymaster), [issues](https://github.com/FutureTense/keymaster/issues))
> First occurred: 12:43:48 AM (4772 occurrences)
> Last logged: 10:48:57 AM
> 
>     [Coordinator] front_door: Code Slot 1: Unable to clear PIN. NotConnected:
>     [Coordinator] front_door: Code Slot 3: Unable to clear PIN. NotConnected:
>     [Coordinator] frontdoor: Code Slot 1: Unable to clear PIN. NotConnected:
>     [Coordinator] frontdoor: Code Slot 3: Unable to clear PIN. NotConnected:
> ```

Are you using the beta or are you using the current release? Because if you're not using the beta, please open a separate issue. That being said, this sounds more like a ZWave connectivity issue than an issue with Keymaster.

**Comment by jleach87** [comment]
@tykeal Yes, running the beta. But I believe your diagnosis was correct. Swapped in a 6ft USB extension cable and issue seems resolved. Thanks a bunch!

**Comment by vamichael** [comment]
I’ve read through some of this, because at this point it’s a little TLDR for me.  While I understand the desire to move logic out of the hands of HA (scripts, automations, helpers, etc.) into something else (what I don’t understand at the moment) for being more “clean”, you still have to have the HA elements call these functions.  So when you build a lock “slot” you have to generate the HTML elements and what happens when a change occurs.  And this happens for every slot.  Sure if you change logic for a certain action you won’t have to change it for every slot. This versus having to re-run the setup routine again to propagate the logic.

**So what are you gaining by doing this?**  Setup?  Resources?  Yeah, things will be smaller but it’s not like the current approach is overwhelming anything.  So please, try to explain what you think the problem is, and how you think it could be resolved.  

But explain it succinctly, like you’re explaining to a bright 20yo that can build a box, install Linux, and get HomeAssistant up and running.  But nothing beyond that.  Don’t assume they know how to use Python, HTML, or other tech, but assume they’re bright enough to understand/learn it.

**Comment by raman325** [comment]
Sorry for the late response here but this is what my lock code manager integration tries to do. Right now I'm having some timing/order of operations issues because we are not relying on HAs templating infrastructure. I'd be happy to accept contributions there if you see some flaws in the design. I will review this PR later but I would advise not doing a complete rearchitecture here. I tried once and it started creating issues that were tough to troubleshoot. The userbase for keymaster is sizable so we should be careful introducing major changes

**Comment by raman325** [comment]
Here's the integration in question: https://github.com/raman325/lock_code_manager

It'll be a lot easier to add other "providers" (zigbee locks, wifi locks, etc) to that one as well if that's of any interest

**Comment by FutureTense** [comment]
> we are not relying on HAs templating infrastructure

Why not, and how are you getting around it?  IIRC templating is used all over the place, especially in the UI.

So you’re doing a fork because you can’t get something to work the way you want it?  If there’s a bug causing the timing issues, it can be fixed.

**Comment by FutureTense** [comment]
- Synchronize multiple locks with a single set of codes
- Optionally use a calendar entity to activate and deactivate a code
- Optionally define a maximum number of uses for a code before the code is disabled

KeyMaster can already do the last two.  I’m not sure what you mean by the first one.   Can you explain how this would work?

**Comment by raman325** [comment]
> Why not, and how are you getting around it?

I wanted as much of it to go through code as possible because it will be lighter and won't create extra entities, just the ones we need. But I built it independently because it's a completely different approach to the syncing logic and has its own challenges (at the moment)

**Comment by FutureTense** [comment]
> because it will be lighter and won't create extra entities,

Is this a problem anywhere now?  I generated 1000 slots once and had no problems.  My machine is a bit more than an RPI, but those are powerful nowadays.  

How are you storing the database?

**Comment by raman325** [comment]
I am not sure which database you are talking about, it's still all Home Assistant under the hood. Anyway, that's a topic we can discuss offline, let's focus this discussion on the PR

**Comment by FutureTense** [comment]
In KeyMaster, the UI *is* the database.  The PIN is stored in the slot’s txtPin (or whatever it’s called) field, for example.  Do you have any screenshots?  Also, I’m still unclear about the sheer n7mbr of elements.  Is it causing problems?

**Comment by Snuffy2** [comment]
So, my goal was to move away from the HA helpers, automations, and scripts and handle all of that internally in Python. It doesn't necessarily reduce the number of entities per lock but ties them all to the keymaster lock and removes all of the automations and scripts. This aligns with how most of the other HA integrations work.

Code-wise it uses the Data Update Coordinator to handle essentially everything. The keymaster entities all get their states from the coordinator and send any changes back to it. The coordinator is not for a single lock but for all keymaster locks so that it can handle child, parent relationships.

I've tried to make this new version function the same as the current keymaster and even built a large migration to move all of the settings from the HA helper entities over to integration entities and remove the HA helpers, automations, and scripts.

The current beta is essentially done as works as intended. I've been running it since around the end of December without any issues. 

I'm a bit disappointed that it sounds like there is no/not much interest in moving forward with this. That was why I had asked if there was interest before starting to build it as I've now put a lot of hours into it. However, it is ultimately your decision.

Let me know if I can answer any more questions.

**Comment by raman325** [comment]
sorry to sound discouraging @Snuffy2 where can I see your changes?

**Comment by firstof9** [comment]
> where can I see your changes?

They're in the beta branch.

**Comment by tykeal** [comment]
@Snuffy2 I like what you've done on the beta it brings a lot of the elegance of Lock Code Manager (LCM) to Keymaster (KM) but keeps the advanced features that I like about KM in place. I'm just personally running into some minor issues with Rental Control (RC) which sits on top of KM because of some design decisions you made (see my comment about keeping "unused" entities as active even if not used).

Migrating the entities configuration into code makes things a lot more clean from the adding, updating, removing perspective of the integration itself. Adding all of the entities to a new KM device is awesome since it automatically causes them to get grouped nicely in the HA UI _and_ it makes it easy for me to detect if we're on the current production KM or the beta without having to do weird things related to version detection. Finally, getting rid of all the automations and moving those to code has significantly improved the lock programming stability in my testing. I haven't had issues on my test instance where the lock gets stuck adding or removing a code with the beta, but I do still semi-regularly get it with the current release of KM. Of course, since i don't have RC fully working with the beta yet I don't have it in place on any of my other systems that see significantly more use so I can't speak fully to if the biggest complaint that folks seem to have with KM is truly fixed with the beta but my more limited testing shows that it probably is.

@FutureTense wrote:
> * Synchronize multiple locks with a single set of codes
> * Optionally use a calendar entity to activate and deactivate a code
> * Optionally define a maximum number of uses for a code before the code is disabled
> 
> KeyMaster can already do the last two. I’m not sure what you mean by the first one. Can you explain how this would work?

KM actually does the first and third of those. The second is is not done via a calendar entity like LCM does it. You're doing it with multiple different datetime entities. For the most generic, you're using "start" and "end" datetime objects to timebox around what is effectively a calendar event, potentially multi-day. This is honestly the _best_ thing for me with RC as I need a way for my users to override the start and end times of the event in question if they need to. If it was an actual calendar entity, it would be significantly harder (see some of my comments below around LCM). The more complex version is using a series of "start" and "end" time only entities that are pointed at specific days of the week. This is also awesome as it's easy to setup a repeating weekly schedule if you have something that always happens on a specific day of the week at a given time (say a cleaner that comes weekly on Wednesday between 10:00 and 14:00). This is doable with a single calendar entity which would support more complex scheduling around a given slot the way LCM does it. The biggest hurdle for KM to move to a this model is migrating current schedules into new calendar entities that would have to get created for the user

Both of these methods time management methods could be collapsed into using a single calendar entity per lock slot the way that LCM does it which is definitely the more elegant solution, but makes my working on RC a lot harder from a usability perspective of a property manager.

@raman325 I like what I see in LCM, but your approach with doing all the management of when a code is active or not based on dedicated calendars is significantly harder to work with from the point of view of RC. My user base is a tiny fraction of that of KM (see the following stats as of the time I wrote this comment) but most of them use RC to drive KM (some don't, but most do) and having to create a dedicated calendar per each slot and then having to come up with a way for users of RC to manage start / end of the event to allow for override of check-in/out times is just a lot more overhead for my integration to deal with when it comes to LCM. Not impossible, but significantly different. If both integrations decided to go with this approach I would definitely adapt, but I would see some really funky code come out of that since I would be having to generate N+1 calendar entities, N being number of slots RC is having to wrangle, and then keeping them in proper sync with the +1, the actual event calendar.

I do love the way that LCM auto generates a dashboard for you. It's not quite the dashboard I need for my systems using RC but it's definitely a great touch for folks that just need a simple lock management setup.

To make KM more extensible to more lock platforms other than ZWave, it would need to adopt _more_ of what LCM is currently doing with a dedicated driver module for ZWave and the ability to write other drivers for other lock systems. This would obviously all be a lot more easy if Home Assistant had a singular lock integration that took care of all of the differing platforms in a simple and easy to use manner that presented everything either code control system needed in a single API.

All in all, there are aspects of both integrations that I really like. I field a lot of support issues for KM because of folks using RC on top of it. I fully expect that I would end up doing the same for LCM if I ever manage to figure out a clean way to support LCM under RC as well but with the differing approaches to slot time management it's harder for me to do so when I've invested as much time into how KM already does stuff. This is partly from my own PRs into KM several years ago when I was first working on RC to get the extra features that I needed in.

Stats:
```
[tykeal@tykes-place]$ get_stats.sh
keymaster:
{
  "total": 1298,
  "versions": {
    "v0.0.99": 840,
    "v0.0.98": 314,
    "v0.0.88": 23,
    "v0.0.85": 41,
    "v0.0.97": 11,
    "v0.0.4": 1,
    "v0.0.87": 18,
    "v0.0.99-b0": 8,
    "v0.0.0": 4,  <-- one of these is me while I'm working on getting RC to work with the beta
    "v0.0.92": 1,
    "v0.0.82": 15,
    "v0.0.80": 2,
    "v0.0.91": 2,
    "v0.0.94": 6,
    "v0.0.77": 4,
    "v0.0.79": 4,
    "v0.0.84": 1,
    "v0.0.96": 1,
    "v0.0.31": 1,
    "v0.0.93": 1
  }
}
lock_code_manager:
{
  "total": 28,
  "versions": {
    "0.5.1": 27,
    "0.0.0": 1   <-- probably @raman325 's test instance
  }
}
rental_control:
{
  "total": 142,
  "versions": {
    "v1.2.0": 59,
    "v1.2.2": 57,  <-- 5 of these are me with all my production HA systems
    "v1.1.2": 12,
    "v1.0.2": 2,
    "v1.0.1": 4,
    "v1.1.1": 2,
    "v0.9.2": 1,
    "v0.9.0": 2,
    "v0.9.1": 2,
    "v0.0.0": 1    <-- my test instance
  }
}
```

**Comment by muddro1** [comment]
I will just say I am super happy with the beta for an updated dashboard, but most of all, not having a ton automatons and helpers cluttering home assistant. They are all great tools, but all the entities from traditional KM was always annoying to deal with.

**Comment by FutureTense** [comment]
> > 
> > KeyMaster can already do the last two. I’m not sure what you mean by the first one. Can you explain how this would work?
> 
> KM actually does the first and third of those. The second is is not done via a calendar entity like LCM does it. You're doing it with multiple different datetime entities. For the most generic, you're using "start" and "end" datetime objects to timebox around what is effectively a calendar event, potentially multi-day. This is honestly the _best_ thing for me with RC as I need a way for my users to override the start and end times of the event in question if they need to. If it was an actual calendar entity, it would be significantly harder (see some of my comments below around LCM). The more complex version is using a series of "start" and "end" time only entities that are pointed at specific days of the week. This is also awesome as it's easy to setup a repeating weekly schedule if you have something that always happens on a specific day of the week at a given time (say a cleaner that comes weekly on Wednesday between 10:00 and 14:00). This is doable with a single calendar entity which would support more complex scheduling around a given slot the way LCM does it. The biggest hurdle for KM to move to a this model is migrating current schedules into new calendar entities that would have to get created for the user

I built the UI the way I did because of its simplicity.   Using checkboxes for MTWRF is a lot easier than trying to define weekdays.  Adding a Calendar entity to HA would be simple.   Just define one for each slot as well as it’s companion “active” Boolean.  cottageBackDoor_1_calEntity, cottageBackDoor_1_calActive

The hard part is finding a UI to populate cottageBackDoor_1_calEntity .  Might be a good candidate for AI.   “Create a calendar entity that’s active weekdays, from 0600-1800”  Imagine if that actually worked?  You could create the craziest schedule and be able to show it’s visual representation as a CalendarEntity,  and store its definition as the string used by the AI.  And of course you could add as many definitions as you want.
cottageBackDoor_1_calEntity - “All Banking Holidays that fall on Monday or Friday” to handle those pesky 3/4 day weekends.  tbh, I like the textual representation better.  But can a Calendar Entity be defined from text?  I sure couldn’t find anything years ago, so I went with the simple approach.

**Comment by justinb186** [comment]
Are these changes going to make it upstream from the beta branch? Or is this abandoned? Having a hard time understanding if this should still be tested or not. Happy to help test on my end if it's still being considered.

**Comment by tykeal** [comment]
You're welcome to use it. In my experience it works better than the current release. The only issue that I'm aware of is actually on me related to Rental Control which is a consumer of Keymaster and I haven't had the focus time to get a new release out that can support both versions.

If you're not using Rental Control at all, then using the beta would be perfectly reasonable.

**Comment by loskinski** [comment]
Just wanted to drop in and say I have been using these changes for a bit now in one of our homes.  WOW is it much better, I hope that these get integrated to main soon!

**Comment by ronaldheft** [comment]
I’m looking forward to the update. I’m about to switch my Yale WiFi modules to Zwave and I’m hoping this update will be out by then. If not, beta it is! 😂

**Comment by sirmeili** [comment]
I just set this up after my lock freaked out and I completely readded all my locks. I'm happy with the beta and the far fewer items in HA. I'll be testing it out (basically only use the auto lock functionality right now), but I'll definitely report back if I have any issues

**Comment by fireboy1919** [comment]
It's been a year since this effort was started.  The automation format that I have on my machine for keymaster is bring deprecated this year.

It's now a maintenance problem as well as a performance problem.

What is currently preventing this from going to master?  Is FutureTense against it?

**Comment by tykeal** [comment]
Well, given that all of the current releases have been off of the beta branch, it would probably be a good idea to merge that into master and stop calling it beta. The beta branch has effectively become the current mainline branch.

There's definitely work to do to fix some issues with the new code base but in general it works significantly better than the old code base and we don't have a bunch of non-connected entities or automations laying around.

**Comment by tykeal** [comment]
As of v0.1.0 this is now released.
